core/src/main/scala/fs2/async/mutable/Signal.scala

```diff
@@ -70,18 +70,23 @@ object Signal {
 
       def discrete: Stream[F, A] = {
         def go(lastA:A, last:Long):Stream[F,A] = {
-          def getNext:F[(A,Long)] = {
+          def getNext:F[(F[(A,Long)],F[Unit])] = {
             F.bind(F.ref[(A,Long)]) { ref =>
-              F.bind(F.modify(state) { case s@(a, l, listen) =>
+              F.map(F.modify(state) { case s@(a, l, listen) =>
                 if (l != last) s
                 else (a, l, listen :+ ref)
               }) { c =>
-                if (c.modified) F.get(ref)
-                else F.pure(c.now._1 -> c.now._2)
+                if (c.modified) {
+                  val cleanup = F.map(F.modify(state) {
+                    case s@(a, l, listen) => if (l != last) s else (a, l, listen.filterNot(_ == ref))
+                  })(_ => ())
```

Comment:
I talked to Pavel about this in Gitter -- we can't use `eq` because `Ref` is an unbounded abstract type member. If we want to be able to use `eq`, we'd need to add a bound `type Ref[A] <: AnyRef`.